src/lib/operators/join_hash.cpp

```diff
@@ -698,55 +723,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
 
       // we need to swap back the inputs, so that the order of the output columns is not harmed
       if (_inputs_swapped) {
-        write_output_chunks(output_chunk, _right_in_table, right, ref_col_right);
+        write_output_chunks(output_chunk, _right_in_table, right_pos_lists_by_column, right, ref_col_right);
 
         // Semi/Anti joins are always swapped but do not need the outer relation
-        if (_mode != JoinMode::Semi && _mode != JoinMode::Anti) {
-          write_output_chunks(output_chunk, _left_in_table, left, ref_col_left);
+        if (output_left_side) {
+          write_output_chunks(output_chunk, _left_in_table, left_pos_lists_by_column, left, ref_col_left);
         }
       } else {
-        write_output_chunks(output_chunk, _left_in_table, left, ref_col_left);
-        write_output_chunks(output_chunk, _right_in_table, right, ref_col_right);
+        write_output_chunks(output_chunk, _left_in_table, left_pos_lists_by_column, left, ref_col_left);
+        write_output_chunks(output_chunk, _right_in_table, right_pos_lists_by_column, right, ref_col_right);
       }
       _output_table->emplace_chunk(std::move(output_chunk));
     }
 
     return _output_table;
   }
 
+  // See usage in _on_execute() for doc.
+  static PosListsByColumn setup_pos_lists_by_column(const std::shared_ptr<const Table> input_table) {
```

Comment:
This gives you a list of all the PosLists for one column, correct? I don't yet understand why this helps.

Reply:
The key is that it returns a shared ptr to a PosList for each column, making it possible to identify columns in the input table that can share output pos lists.